### PR TITLE
ver 1.9.0

### PR DESCRIPTION
**Pull request by mbdaryanto**
Add Xsolla IAP

### CODE REVIEW

File: src/auth/xsolla.ts [hunk 1]
```diff
@@ -37,16 +40,28 @@ export function getXsollaWidget() {
...
+        xsollaWidget.once(xsollaWidget.events.Close, function (e: any) {
             console.log('user has closed the widget');
             el.style.display = 'none';
-            resolve();
+            reject(e);
+        });
+        xsollaWidget.once(xsollaWidget.events.LoginSuccess, function (event: any) {
+            console.log('user logged in successfully');
+            el.style.display = 'none';
+            console.log({event});
```
Comment (briandeheus) [inline src/auth/xsolla.ts:57]: Can you remove this?

File: src/auth/xsolla.ts [hunk 2]
```diff
@@ -37,16 +40,28 @@ export function getXsollaWidget() {
...
+            resolve(event.params.token);
+        });
+        xsollaWidget.once(xsollaWidget.events.SignupSuccess, function (event: any) {
+            console.log('user signed up successfully');
+            el.style.display = 'none';
+            console.log({event});
```
Comment (briandeheus) [inline src/auth/xsolla.ts:63]: Remove this.

File: src/auth/xsolla.ts [hunk 3]
```diff
@@ -37,16 +40,28 @@ export function getXsollaWidget() {
...
+            reject(e);
+        });
+        xsollaWidget.once(xsollaWidget.events.LoginSuccess, function (event: any) {
+            console.log('user logged in successfully');
+            el.style.display = 'none';
+            console.log({event});
+            resolve(event.params.token);
+        });
+        xsollaWidget.once(xsollaWidget.events.SignupSuccess, function (event: any) {
+            console.log('user signed up successfully');
```
Comment (briandeheus) [inline src/auth/xsolla.ts:61]: Should we use the wortal debug logger?

File: src/iap/impl/iap-xsolla.ts [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+
+        return items.map(item => ({
+            title: item.name,
+            productID: item.sku,
+            description: item.description,
+            imageURI: item.image_url,
+            price: `${item.price.amount} ${item.price.currency}`,
+            priceCurrencyCode: item.price.currency,
+            priceAmount: Number(item.price.amount),
+            // _internalData: item,
```
Comment (briandeheus) [inline src/iap/impl/iap-xsolla.ts:97]: Remove commented out code.

File: src/iap/impl/iap-xsolla.ts [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+        }));
+    }
+
+    protected async getPurchasesAsyncImpl(): Promise<Purchase[]> {
+        // return Promise.reject(notSupported(undefined, WORTAL_API.IAP_GET_PURCHASES_ASYNC, API_URL.IAP_GET_PURCHASES_ASYNC));
```
Comment (briandeheus) [inline src/iap/impl/iap-xsolla.ts:102]: Remove commented out code.

File: src/iap/impl/iap-xsolla.ts [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+    protected getSubscriptionsAsyncImpl(): Promise<Subscription[]> {
+        return Promise.reject(notSupported(undefined, WORTAL_API.IAP_GET_SUBSCRIPTIONS_ASYNC, API_URL.IAP_GET_SUBSCRIPTIONS_ASYNC));
+    }
+
+    protected isEnabledImpl(): boolean {
+        return true;
+    }
+
+    protected async makePurchaseAsyncImpl(purchase: PurchaseConfig): Promise<Purchase> {
+        // return Promise.reject(notSupported(undefined, WORTAL_API.IAP_MAKE_PURCHASE_ASYNC, API_URL.IAP_MAKE_PURCHASE_ASYNC));
```
Comment (briandeheus) [inline src/iap/impl/iap-xsolla.ts:143]: Remove commented out code.

File: src/iap/impl/iap-xsolla.ts [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+        Wortal._log.debug(`IAP initialized for ${Wortal._internalPlatform} platform.`);
+    }
+}
+
+const { XSOLLA_STORE } = API_ENDPOINTS;
+const STORE_BASE_URL = `${XSOLLA_STORE}/v2/project`;
+
+/**
+ * fetchVirtualItems from xsolla store
+ * @param param0
```
Comment (briandeheus) [inline src/iap/impl/iap-xsolla.ts:186]: param0 doesn't make sense.

File: src/player/impl/player-crazygames.ts [hunk 8]
```diff
@@ -29,9 +32,9 @@ export class PlayerCrazyGames extends PlayerBase {
         return Promise.reject(notSupported(undefined, WORTAL_API.PLAYER_GET_CONNECTED_PLAYERS_ASYNC, API_URL.PLAYER_GET_CONNECTED_PLAYERS_ASYNC));
     }
 
-    protected getDataAsyncImpl(keys: string[]): Promise<any> {
-        return this.defaultGetDataAsyncImpl(keys);
-    }
+    // protected getDataAsyncImpl(keys: string[]): Promise<any> {
```
Comment (briandeheus) [inline src/player/impl/player-crazygames.ts:35]: Remove commented out code.

File: src/player/impl/player-crazygames.ts [hunk 9]
```diff
@@ -51,12 +54,96 @@ export class PlayerCrazyGames extends PlayerBase {
...
             };
 
-            Wortal._internalPlatformSDK.user.getUserToken(callback);
+            (Wortal._internalPlatformSDK as CrazyGamesSDK).user.getUserToken(callback);
         });
     }
 
-    protected setDataAsyncImpl(data: Record<string, unknown>): Promise<void> {
-        return this.defaultSetDataAsyncImpl(data);
+    // protected setDataAsyncImpl(data: Record<string, unknown>): Promise<void> {
```
Comment (briandeheus) [inline src/player/impl/player-crazygames.ts:61]: Remove commented out code.

File: src/auth/xsolla.ts (hunk 1 above)
Comment (mbdaryanto) [reply src/auth/xsolla.ts:57]: I will replace console.log with Wortal._log.debug

File: src/auth/xsolla.ts (hunk 3 above)
Comment (mbdaryanto) [reply src/auth/xsolla.ts:61]: ok

File: src/auth/xsolla.ts (hunk 2 above)
Comment (mbdaryanto) [reply src/auth/xsolla.ts:63]: I replace it with Wortal._log.debug

File: src/iap/impl/iap-xsolla.ts (hunk 4 above)
Comment (mbdaryanto) [reply src/iap/impl/iap-xsolla.ts:97]: ok

File: src/iap/impl/iap-xsolla.ts (hunk 6 above)
Comment (mbdaryanto) [reply src/iap/impl/iap-xsolla.ts:143]: ok

File: src/iap/impl/iap-xsolla.ts (hunk 5 above)
Comment (mbdaryanto) [reply src/iap/impl/iap-xsolla.ts:102]: ok

File: src/iap/impl/iap-xsolla.ts (hunk 7 above)
Comment (mbdaryanto) [reply src/iap/impl/iap-xsolla.ts:186]: I'll update this

File: src/player/impl/player-crazygames.ts (hunk 8 above)
Comment (mbdaryanto) [reply src/player/impl/player-crazygames.ts:35]: ok

File: src/player/impl/player-crazygames.ts (hunk 9 above)
Comment (mbdaryanto) [reply src/player/impl/player-crazygames.ts:61]: ok

**Pull request merged**